### ISSUE TITLE
Wrong deprecation warning on using task outputs in another task with no explicit task dependency 

### ISSUE THREAD

**Issue by breskeby** [gradle/gradle#16160]
When declaring a filtered fileTree as input for a task, gradle emits a deprecation warning about using output from another task even though this output is filtered out explicitly in the tasks inputs declaration.

### Steps to Reproduce
 
With the sample project attached run:

In the script we filter out the `build` folder explicitly from the inputs of myTask (which you can verify by checking the output of the `myTask`.

But running 

```
./gradlew clean compileJava myTask
``` 

still emits this deprecation warning:

```
> Task :myTask
Validation failed for task ':myTask', disabling optimizations:
  - Task ':myTask' uses the output of task ':compileJava', without declaring an explicit dependency (using Task.dependsOn() or Task.mustRunAfter()) or an implicit dependency (declaring task ':compileJava' as an input). This can lead to incorrect results being produced, depending on what order the tasks are executed.
Task ':myTask' uses the output of task ':compileJava', without declaring an explicit dependency (using Task.dependsOn() or Task.mustRunAfter()) or an implicit dependency (declaring task ':compileJava' as an input). This can lead to incorrect results being produced, depending on what order the tasks are executed. This behaviour has been deprecated and is scheduled to be removed in Gradle 7.0. Execution optimizations are disabled due to the failed validation. See https://docs.gradle.org/7.0-milestone-1/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
```

[overlappinginputs.zip](https://github.com/gradle/gradle/files/5983641/overlappinginputs.zip)

### Your Environment

- Running with Gradle 7.0-milestone-1

Build scan URL:
https://gradle.com/s/pz3rqikhptn5g

**Comment by wolfs** [comment]
Hmm... If you use `exclude 'build/**'` instead of `exclude 'build'` then it works. So seems like we are interpreting the excludes not consistent with what file tree is doing.

**Comment by breskeby** [comment]
@wolfs thanks for looking into this. interesting. This at least unblocks us from migrating

**Comment by wolfs** [comment]
@breskeby I have a fix for this in #16174. Thank you for bringing the bug to my attention!

As a clarification: This new warning about missing task dependencies is a new deprecation in 7.0 and will fail your build in 8.0. So this deprecation won't block you from upgrading to 8.0.

**Comment by breskeby** [comment]
Thanks. Yes understood. But we currently fail for any deprecation warning. So we want this sorted as part of the migration
